File: pyhessref/nuc_repl.py
import numpy as np
from pyscf import gto

from pyhessref.hess_trait_restricted import RHessCoreAPI


def get_nuc_repl_hess(mol: gto.Mole) -> np.ndarray:
    """Hessian contribution from nuclear repulsion.

    Parameters
    ----------
    mol : gto.Mole
        The molecule object.

    Returns
    -------
    de_nuc : np.ndarray
        The nuclear repulsion Hessian, shape [natm, natm, 3, 3].
    """
    natm = mol.natm
    de_nuc = np.zeros([natm, natm, 3, 3])

    qs = np.asarray([mol.atom_charge(i) for i in range(natm)])
    rs = np.asarray([mol.atom_coord(i) for i in range(natm)])
    for i in range(natm):
        r12 = rs[i] - rs  # shape (natm, 3)
        s12 = np.sqrt(np.sum(r12 * r12, axis=1))  # einsum: 'ki,ki->k'
        s12[i] = np.inf  # avoid division by zero
        tmp1 = qs[i] * qs / s12**3  # shape [natm]
        prefactor = -3 * qs[i] * qs / s12**5  # shape [natm]
        tmp2 = prefactor[:, None, None] * r12[:, :, None] * r12[:, None, :]

        # Diagonal block h[i,i]
        de_nuc[i, i, 0, 0] = de_nuc[i, i, 1, 1] = de_nuc[i, i, 2, 2] = -tmp1.sum()
        de_nuc[i, i] -= np.sum(tmp2, axis=0)  # einsum: 'kij->ij'

        # Off-diagonal blocks h[i,:] for all k
        de_nuc[i, :, 0, 0] += tmp1
        de_nuc[i, :, 1, 1] += tmp1
        de_nuc[i, :, 2, 2] += tmp1
        de_nuc[i, :] += tmp2
    return de_nuc


class HessNucRepl(RHessCoreAPI):
    """Hessian contribution from nuclear repulsion."""

    def __init__(self, mol: gto.Mole):
        self.mol = mol

    def make_skeleton_hess(self, *args, **kwargs) -> np.ndarray:
        return get_nuc_repl_hess(self.mol)

    def generator_deriv1(self) -> callable:
        return None

File: src/hessian/nuc_repl.rs
use crate::prelude::*;

/// Hessian contribution from nuclear repulsion.
///
/// # Parameters
///
/// - `mol` : [`CInt`]. The molecule object.
/// - `device` : [`DeviceTsr`]. The device on which the returned tensor is allocated.
///
/// # Returns
/// -------
/// - `de_nuc` : shape [3, 3, natm, natm]. The nuclear repulsion Hessian.
pub fn get_nuc_repl_hess(mol: &CInt, device: &DeviceTsr) -> Tsr {
    let natm = mol.natm();

    // de_nuc: shape [3, 3, natm, natm]. The nuclear repulsion Hessian.
    let mut de_nuc: Tsr = rt::zeros(([3, 3, natm, natm], device));
    // qs: shape [natm]. The atomic charges.
    let qs = rt::asarray((mol.atom_charges(), device));
    // rs: shape [3, natm]. The atomic coordinates.
    let rs = rt::asarray((mol.atom_coords(), device)).into_unpack_array(0);

    for A in 0..natm {
        // r12: shape [3, natm]. The vector from atom A to other atoms.
        let r12 = rs.i((.., A)) - &rs;
        // s12: shape [natm]. The distance from atom A to other atoms.
        //      this value will be divided, so we set zero distance to inf to avoid NaN.
        let mut s12 = r12.l2_norm_axes(0);
        s12[[A]] = f64::INFINITY;

        // tmp1: shape [natm]
        let tmp1 = qs[[A]] * &qs / s12.pow(3);
        // prefactor: shape [natm]
        let prefactor = -3.0 * qs[[A]] * &qs / s12.pow(5);
        // tmp2: shape [3, 3, natm]
        let tmp2 = prefactor.i((None, None, ..)) * r12.i((.., None, ..)) * r12.i((None, .., ..));

        // diagonal block
        let tmp1_sum = tmp1.sum(); // scalar
        let tmp2_sum = tmp2.sum_axes(-1); // shape [3, 3]
        de_nuc[[0, 0, A, A]] -= tmp1_sum;
        de_nuc[[1, 1, A, A]] -= tmp1_sum;
        de_nuc[[2, 2, A, A]] -= tmp1_sum;
        *&mut de_nuc.i_mut((.., .., A, A)) -= tmp2_sum;

        // off-diagonal blocks
        *&mut de_nuc.i_mut((0, 0, A, ..)) += &tmp1;
        *&mut de_nuc.i_mut((1, 1, A, ..)) += &tmp1;
        *&mut de_nuc.i_mut((2, 2, A, ..)) += &tmp1;
        *&mut de_nuc.i_mut((.., .., A, ..)) += &tmp2;
    }

    de_nuc
}

/// Hessian contribution from nuclear repulsion.
pub struct HessNucRepl {
    pub mol: CInt,
}

impl HessNucRepl {
    pub fn new(mol: &CInt) -> Self {
        Self { mol: mol.clone() }
    }
}

impl RHessCoreAPI for HessNucRepl {
    fn make_skeleton_hess(&mut self, mo_coeff: TsrView, _mo_occ: TsrView) -> Tsr {
        get_nuc_repl_hess(&self.mol, mo_coeff.device())
    }

    fn generator_deriv1(&self) -> Option<Box<dyn FnMut(usize) -> Tsr>> {
        None
    }
}
